eth-contracts/contracts/DelegateManagerV2.sol

```diff
@@ -0,0 +1,1374 @@
+        require(
+            block.number >= removeDelegatorRequests[_serviceProvider][_delegator],
+            "DelegateManager: Lockup must be expired"
+        );
+
+        // Enforce evaluation window for request
+        require(
+            block.number < removeDelegatorRequests[_serviceProvider][_delegator] + removeDelegatorEvalDuration,
+            "DelegateManager: RemoveDelegator evaluation window expired"
+        );
+
+        uint256 unstakeAmount = delegateInfo[_delegator][_serviceProvider];
+        // Unstake on behalf of target service provider
+        Staking(stakingAddress).undelegateStakeFor(
+            _serviceProvider,
+            _delegator,
+            unstakeAmount
+        );
+        // Update total delegated for SP
+        // totalServiceProviderDelegatedStake - total amount delegated to service provider
+        // totalStakedForSpFromDelegator - amount staked from this delegator to targeted service provider
+        _updateDelegatorStake(
+            _delegator,
+            _serviceProvider,
+            spDelegateInfo[_serviceProvider].totalDelegatedStake.sub(unstakeAmount),
+            delegateInfo[_delegator][_serviceProvider].sub(unstakeAmount),
+            delegatorTotalStake[_delegator].sub(unstakeAmount)
+        );
+
+        if (
+            _undelegateRequestIsPending(_delegator) &&
+            undelegateRequests[_delegator].serviceProvider == _serviceProvider
+        ) {
+            // Remove pending request information
+            _updateServiceProviderLockupAmount(
+                _serviceProvider,
+                spDelegateInfo[_serviceProvider].totalLockedUpStake.sub(undelegateRequests[_delegator].amount)
+            );
+            _resetUndelegateStakeRequest(_delegator);
+        }
+
+        // Remove from list of delegators
+        _removeFromDelegatorsList(_serviceProvider, _delegator);
+
+        // Reset lockup expiry
+        removeDelegatorRequests[_serviceProvider][_delegator] = 0;
+        emit RemoveDelegatorRequestEvaluated(_serviceProvider, _delegator, unstakeAmount);
+    }
+
+    /**
+     * TODO consider naming ServiceProvider instead of SP
+     * note - does not enforce _spMinDelegationAmount >= minDelegationAmount since not necessary
+            delegateStake() and undelegateStake() always take the max of both already
+     */
+    function updateSPMinDelegationAmount(
+        address _serviceProvider,
+        uint256 _spMinDelegationAmount
+    ) external {
+        _requireIsInitialized();
+
+        require(
+            msg.sender == _serviceProvider || msg.sender == governanceAddress,
+            ERROR_ONLY_SP_GOVERNANCE
+        );
+
+        // Ensure valid SP
+        // TODO not sure if needed
+        // ideally would like to access `SPFactory.serviceProviderAddressToId` but it is not externally accessible
+        // using numEndpoints >= 1 as an invariant is dicey
+        (,,, uint256 numEndpoints,,) = (
```

Comment:
resolved afk, will keep in

Reply:
the reason `getServiceProviderIdsFromAddress` doesn't work is bc it requires both address and type, since IDs are only unique per type
delegateManager has no knowledge of type and it doesn't make any sense to iterate over all possible types in order to make that call

i would have loved to have a better validation but this is the best we could think of. i agree its not great but it feels pretty safe

Reply:
im good w it